clicommand/pipeline_eval.go

```diff
@@ -0,0 +1,217 @@
+
+			paths := []string{
+				"buildkite.js",
+				filepath.FromSlash(".buildkite/buildkite.js"),
+				filepath.FromSlash("buildkite/buildkite.js"),
+			}
+
+			// Collect all the files that exist
+			exists := []string{}
+			for _, path := range paths {
+				if _, err := os.Stat(path); err == nil {
+					exists = append(exists, path)
+				}
+			}
+
+			// If more than 1 of the config files exist, throw an
+			// error. There can only be one!!
+			if len(exists) > 1 {
+				l.Fatal("Found multiple configuration files: %s. Please only have 1 configuration file present.", strings.Join(exists, ", "))
+			} else if len(exists) == 0 {
+				l.Fatal("Could not find a default pipeline configuration file. See `buildkite-agent pipeline upload --help` for more information.")
+			}
+
+			found := exists[0]
+
+			l.Info("Found config file \"%s\"", found)
+
+			// Read the default file
+			filename = path.Base(found)
+			input, err = os.ReadFile(found)
+			if err != nil {
+				l.Fatal("Failed to read file \"%s\" (%s)", found, err)
+			}
+		}
+
+		if err := evalJS(filename, input, c.App.Writer); err != nil {
+			panic(err)
+		}
+		return nil
+	},
+}
+
+func evalJS(filename string, input []byte, output io.Writer) error {
+	runtime := goja.New()
+	runtime.SetFieldNameMapper(goja.TagFieldNameMapper("json", true))
+
+	// Add support for require() CommonJS modules.
+	// require("buildkite/*") is handled by embedded resources/node_modules/buildkite/* filesystem.
+	// Other paths are loaded from the host filesystem.
+	registry := require.NewRegistry(
+		require.WithLoader(func(name string) ([]byte, error) {
+			if !strings.HasPrefix(name, "node_modules/buildkite/") {
+				return require.DefaultSourceLoader(name)
+			}
+			res := resources.FS
+			data, err := res.ReadFile(name)
+			if errors.Is(err, fs.ErrNotExist) {
+				return nil, require.ModuleFileDoesNotExistError
+			} else if err != nil {
+				return nil, err
+			}
+			return data, nil
+		}),
+	)
+	registry.Enable(runtime)
+
+	// Add basic utilities
+	console.Enable(runtime) // console.log()
+	process.Enable(runtime) // process.env()
+
+	// provide plugin() as a native module (implemented in Go)
+	registry.RegisterNativeModule("buildkite/plugin", func(runtime *goja.Runtime, module *goja.Object) {
+		module.Set("exports", func(call goja.FunctionCall) goja.Value {
+			name := call.Argument(0)
+			ref := call.Argument(1)
+			config := call.Argument(2)
+			plugin := runtime.NewObject()
+			plugin.Set(name.String()+"#"+ref.String(), config)
+			return plugin
+		})
+	})
+
+	// provide assignable module.exports for Pipeline result
+	rootModule := runtime.NewObject()
+	rootModule.Set("exports", runtime.NewObject())
+	err := runtime.Set("module", rootModule)
+	if err != nil {
+		return err
+	}
+
+	if filename == "" {
+		filename = "(stdin)"
+	}
+
+	v, err := runtime.RunScript(filename, string(input))
+	if err != nil {
+		panic(err)
+	}
+
+	y, err := yaml.Marshal(v.Export())
+	if err != nil {
+		return err
+	}
```

Comment:
it might be cool to evaluate the generated pipeline.yml against the [pipeline json schema](https://github.com/buildkite/pipeline-schema) here to safeguard against generating a wonky pipeline

🤔 i wonder if we should do that on pipeline upload as well, even if we just output warnings

Reply:
Yeah that's be nice on upload too!

Reply:
Oh, hell yeah! If we can get this thing talking to npm, maybe it makes sense to keep the schema in a package. If we kept it in sync, then our pipeline editor in the UI could validate against it too.